pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,127 @@
+    cost function first derivatives with respect to the variational parameters can be cast into
+    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
+    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
+    found in https://arxiv.org/abs/2107.12390.
+
+    Args:
+        frequencies (tuple[int or float]): the tuple of eigenvalue frequencies. Eigenvalue
+            frequencies are defined as the unique positive differences obtained from a set of
+            eigenvalues.
+        shifts (tuple[int or float]): the tuple of shift values. If unspecified, equidistant
+            shifts are assumed. If supplied, the length of this tuple should match the number of
+            given frequencies.
+
+    Returns:
+         tuple: a tuple of one nested list describing the gradient recipe
+         for the parameter-shift method.
+         This is a tuple with one nested list per operation parameter. For
+         parameter :math:`\phi_k`, the nested list contains elements of the form
+         :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+         term, resulting in a gradient recipe of
+
+            .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+
+    Raises:
+        ValueError: if ``frequencies`` is not a list of unique positive values, or if ``shifts``
+            (if specified) is not a list of unique values the same length as ``frequencies``.
+
+    **Examples**
+
+    An example of obtaining the frequencies from a set of unique eigenvalues and obtaining the
+    parameter shift rule:
+
+    >>> unique_eigenvals = sorted([1, -1, 0])
+    >>> from itertools import combinations
+    >>> frequencies = {abs(i - j) for i, j in combinations(unique_eigenvals, 2)}
+    >>> get_shift_rule(tuple(frequencies))
+    ([[0.8535533905932737, 1, 0.7853981633974483], [-0.14644660940672624, 1, 2.356194490192345],
+    [-0.8535533905932737, 1, -0.7853981633974483], [0.14644660940672624, 1, -2.356194490192345]],)
+
+    An example with explicitly specified shift values:
+
+    >>> frequencies = (1, 2, 4)
+    >>> shifts = (np.pi / 3, 2 * np.pi / 3, np.pi / 4)
+    >>> get_shift_rule(frequencies, shifts)
+    ([[-2.0907702751760278, 1, 1.0471975511965976], [0.2186308015824754, 1, 2.0943951023931953],
+    [3.0000000000000004, 1, 0.7853981633974483], [2.0907702751760278, 1, -1.0471975511965976],
+    [-0.2186308015824754, 1, -2.0943951023931953], [-3.0000000000000004, 1, -0.7853981633974483]],)
+    """
+
+    n_freqs = len(frequencies)
+
+    if len(set(frequencies)) != n_freqs or any(freq <= 0 for freq in frequencies):
+        raise ValueError(
+            "Expected frequencies to be a list of unique positive values, instead got {}.".format(
+                frequencies
+            )
+        )
+
+    frequencies = qml.math.sort(qml.math.stack(frequencies))

```

Comment:
If we first do the sorting and then perform the above test by checking whether the first frequency is negative, we save cost in case no `ValueError` is raised (and the additional cost in case that there _is_ an `ValueError` raised is not super-large and less performance-relevant I'd say)